pkg/controller/constrainttemplate/constrainttemplate_controller_test.go

```diff
@@ -183,6 +257,135 @@ func TestReconcile(t *testing.T) {
+		labels := map[string]string{
+			"gatekeeper.sh/use-vap": "yes",
+		}
+		constraintTemplate := makeReconcileConstraintTemplateForVap(suffix, labels)
+		t.Cleanup(testutils.DeleteObjectAndConfirm(ctx, t, c, expectedCRD(suffix)))
+		testutils.CreateThenCleanup(ctx, t, c, constraintTemplate)
+
+		err = retry.OnError(testutils.ConstantRetry, func(err error) bool {
+			return true
+		}, func() error {
+			// check if vap resource exists now
+			vap := &admissionregistrationv1beta1.ValidatingAdmissionPolicy{}
+			vapName := fmt.Sprintf("gatekeeper-%s", denyall+strings.ToLower(suffix))
+			if err := c.Get(ctx, types.NamespacedName{Name: vapName}, vap); err != nil {
+				return err
+			}
+			return nil
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+	})
+
+	t.Run("Vap should not be created", func(t *testing.T) {
+		suffix := "VapShouldNotBeCreated"
+
+		logger.Info("Running test: Vap should not be created")
+		labels := map[string]string{
+			"gatekeeper.sh/use-vap": "no",
+		}
+		constraintTemplate := makeReconcileConstraintTemplateForVap(suffix, labels)
+		t.Cleanup(testutils.DeleteObjectAndConfirm(ctx, t, c, expectedCRD(suffix)))
+		testutils.CreateThenCleanup(ctx, t, c, constraintTemplate)
+
+		err = retry.OnError(testutils.ConstantRetry, func(err error) bool {
+			return true
+		}, func() error {
+			// check if vap resource exists now
+			vap := &admissionregistrationv1beta1.ValidatingAdmissionPolicy{}
+			vapName := fmt.Sprintf("gatekeeper-%s", denyall+strings.ToLower(suffix))
+			if err := c.Get(ctx, types.NamespacedName{Name: vapName}, vap); err != nil {
+				if !apierrors.IsNotFound(err) {
+					return err
+				}
+				return nil
+			}
+			return fmt.Errorf("should result in error, vap not found")
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+	})
+
+	t.Run("Vap should not be created without label", func(t *testing.T) {
+		suffix := "VapShouldNotBeCreatedWithoutLabel"
+
+		logger.Info("Running test: Vap should not be created without label")
+		labels := map[string]string{}
+		constraintTemplate := makeReconcileConstraintTemplateForVap(suffix, labels)
+		t.Cleanup(testutils.DeleteObjectAndConfirm(ctx, t, c, expectedCRD(suffix)))
+		testutils.CreateThenCleanup(ctx, t, c, constraintTemplate)
+
+		err = retry.OnError(testutils.ConstantRetry, func(err error) bool {
+			return true
+		}, func() error {
+			// check if vap resource exists now
+			vap := &admissionregistrationv1beta1.ValidatingAdmissionPolicy{}
+			vapName := fmt.Sprintf("gatekeeper-%s", denyall+strings.ToLower(suffix))
+			if err := c.Get(ctx, types.NamespacedName{Name: vapName}, vap); err != nil {
+				if !apierrors.IsNotFound(err) {
+					return err
+				}
+				return nil
+			}
+			return fmt.Errorf("should result in error, vap not found")
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+	})
+	// TODO(ritazh): uncomment this test after the fix for https://github.com/kubernetes/kubernetes/issues/122658 makes its way to a k8s release
```

Comment:
should we do custom build for this? we have done a custom kind node image in the past, but this'll be for envtest i think

Reply:
https://github.com/kubernetes-sigs/controller-runtime/tree/main/tools/setup-envtest

Reply:
It looks like you can only specify a supported k8s version, not a commit that has not landed as part of any release.